File: backend/src/main/kotlin/com/dumpdiary/backend/security/TokenService.kt
package com.dumpdiary.backend.security

import com.dumpdiary.backend.repository.InMemoryStore
import com.dumpdiary.backend.repository.RefreshTokenRecord
import io.ktor.server.auth.Principal
import java.util.UUID

data class AuthenticatedUser(val userId: String) : Principal

class TokenService(private val store: InMemoryStore) {
    fun issueAccessToken(userId: String): String {
        val token = "access-${UUID.randomUUID()}"
        store.putAccessToken(token, userId)
        return token
    }

    fun issueRefreshToken(userId: String): String {
        val token = "refresh-${UUID.randomUUID()}"
        store.putRefreshToken(token, RefreshTokenRecord(userId = userId, token = token))
        return token
    }

    fun resolveAccessToken(token: String): String? = store.resolveAccessToken(token)

    fun refresh(refreshToken: String): String? = store.getRefreshToken(refreshToken)?.userId?.let(::issueAccessToken)

    fun revoke(accessToken: String?, refreshToken: String?) {
        store.removeAccessToken(accessToken)
        store.removeRefreshToken(refreshToken)
    }
}

File: backend-node/src/token-service.js
import { randomUUID } from "node:crypto";

export class TokenService {
  constructor(store) {
    this.store = store;
  }

  issueAccessToken(userId) {
    const token = `access-${randomUUID()}`;
    this.store.putAccessToken(token, userId);
    return token;
  }

  issueRefreshToken(userId) {
    const token = `refresh-${randomUUID()}`;
    this.store.putRefreshToken(token, { userId, token });
    return token;
  }

  resolveAccessToken(token) {
    return this.store.resolveAccessToken(token);
  }

  refresh(refreshToken) {
    const record = this.store.getRefreshToken(refreshToken);
    return record ? this.issueAccessToken(record.userId) : null;
  }

  revoke(accessToken, refreshToken) {
    this.store.removeAccessToken(accessToken);
    this.store.removeRefreshToken(refreshToken);
  }
}
